reactor-core/src/main/java/reactor/core/publisher/Traces.java

```diff
@@ -97,80 +95,164 @@ static String extractOperatorAssemblyInformation(String source) {
+	 * from the assembly stack trace
 	 */
 	static String[] extractOperatorAssemblyInformationParts(String source) {
-		String[] uncleanTraces = source.split("\n");
-		final List<String> traces = Stream.of(uncleanTraces)
-		                                  .map(String::trim)
-		                                  .filter(s -> !s.isEmpty())
-		                                  .collect(Collectors.toList());
+		Iterator<Substring> traces = trimmedNonemptyLines(source);
 
-		if (traces.isEmpty()) {
+		if (!traces.hasNext()) {
 			return new String[0];
 		}
 
-		int i = 0;
-		while (i < traces.size() && !isUserCode(traces.get(i))) {
-			i++;
+		Substring prevLine = null;
+		Substring currentLine = traces.next();
+
+		if (currentLine.isUserCode()) {
+			// No line is a Reactor API line.
+			return new String[]{currentLine.toString()};
 		}
 
-		String apiLine;
-		String userCodeLine;
-		if (i == 0) {
-			//no line was a reactor API line
-			apiLine = "";
-			userCodeLine = traces.get(0);
+		while (traces.hasNext()) {
+			prevLine = currentLine;
+			currentLine = traces.next();
+
+			if (currentLine.isUserCode()) {
+				// Currently on user code line, previous one is API. Attempt to create something in the form
+				// "Flux.map ⇢ user.code.Class.method(Class.java:123)".
+				return new String[]{
+					prevLine.withoutPublisherPackagePrefix().withoutLocationSuffix().toString(),
+					"at " + currentLine};
+			}
 		}
-		else if (i == traces.size()) {
-			//we skipped ALL lines, meaning they're all reactor API lines. We'll fully display the last one
-			apiLine = "";
-			userCodeLine = traces.get(i-1).replaceFirst("reactor.core.publisher.", "");
+
+		// We skipped ALL lines, meaning they're all Reactor API lines. We'll fully display the last
+		// one.
+		return new String[]{currentLine.withoutPublisherPackagePrefix().toString()};
+	}
+
+	/**
+	 * Returns an iterator over all trimmed non-empty lines in the given source string.
+	 *
+	 * @implNote This implementation attempts to minimize allocations.
+	 */
+	private static Iterator<Substring> trimmedNonemptyLines(String source) {
+		return new Iterator<Substring>() {
+			private int index = 0;
+			@Nullable
+			private Substring next = getNextLine();
+
+			@Override
+			public boolean hasNext() {
+				return next != null;
+			}
+
+			@Override
+			public Substring next() {
+				Substring current = next;
+				if (current == null) {
+					throw new NoSuchElementException();
+				}
+				next = getNextLine();
+				return current;
+			}
+
+			@Nullable
+			private Substring getNextLine() {
+				while (index < source.length()) {
+					int end = source.indexOf('\n', index);
+					if (end == -1) {
+						end = source.length();
+					}
+					Substring line = new Substring(source, index, end).trim();
+					index = end + 1;
+					if (!line.isEmpty()) {
+						return line;
+					}
+				}
+				return null;
+			}
+		};
+	}
+
+	// XXX: Explain.
+	private static final class Substring {
```

Comment:
Consider renaming it to `StringView`. This would communicate to the reader that we are only wrapping the underlying `String` and not making any copies of it.

Reply:
Or even `StackLineView`, since it exposes methods specific to the lines in the stack trace.